## 📍 Agenda 📍
<br>

- [Documentação oficial - Numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)
- [Exercícios](#dois)

### Exercícios

1. Importe numpy como 'np' e imprima o número da versão instalada no ambiente (local ou colab).

In [2]:
import numpy as np
print(f"a versão do numpy instalada é: {np.__version__}")


a versão do numpy instalada é: 1.23.4


2. Crie um *array* 1D com números de 0 a 9 e imprima o resultado.

In [46]:
a1d = np.arange(0,10)
a1d

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

3. Crie uma matriz **booleana** de 3x3 e preencha os seus valores com 'True'. Imprima o resultado.

**Dica:** Verifique o uso da função **full** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html).

In [48]:
#e = np.zeros((3,3))
#boleana = e<1
#boleana

from numpy import True_


#e = np.full((3, 3), True) #ele coloca a dimensão e o valor que vc deseja preencher
#e

e=np.zeros((3,3))
e[e==0]= True_
e


array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

4. Crie um *array* 1D como números inteiros (50 elementos) de 0 a 1000 e **extraia** apenas os números pares. Imprima o resultado.

In [27]:
nv = np.random.randint(0,1000,50) 
pares = nv[nv%2==0]
pares


array([584, 764, 320, 518, 726, 938,  20, 980, 888, 216, 976, 822, 550,
       486, 398,  16, 530,  12, 460, 162,  84, 984, 342, 658, 470, 918,
       178, 928, 110])

5. Usando o *array* criando no exercício anterior, **substitua** todos os números ímpares maiores que 50 por -1 no *array* original. 

In [35]:
#ajustar o resultado 
a2 = nv[nv>50]
np.where(a2 % 2 == 0, a2, -1)




array([ -1, 584, 764, 320, 518, 726, 938,  -1,  -1, 980, 888,  -1,  -1,
       216, 976, 822,  -1,  -1,  -1, 550,  -1, 486, 398,  -1,  -1,  -1,
        -1, 530, 460, 162,  84,  -1, 984,  -1,  -1, 342, 658, 470,  -1,
       918, 178, 928, 110,  -1])

6. Usando o *array* criando no exercício anterior, **substitua** todos os números ímpares com o valor -1 **sem** alterar o *array* original.

In [36]:
np.where(nv % 2 == 0, nv, -1) #o where altera uma cópia do aray original, ele não modifica

array([ -1,  -1, 584,  -1, 764, 320, 518, 726, 938,  20,  -1,  -1, 980,
       888,  -1,  -1, 216, 976, 822,  -1,  -1,  -1, 550,  -1, 486, 398,
        -1,  -1,  -1,  -1,  16, 530,  12, 460, 162,  84,  -1, 984,  -1,
        -1,  -1, 342, 658, 470,  -1, 918, 178, 928, 110,  -1])

7. Crie um *array* 1D com 100 valores de ponto flutuante (**float**), converta-o em uma matriz 2D com 10 linhas e 10 colunas.

In [41]:
matriz = np.random.rand(1, 100)
matriz.reshape((10, 10))

array([[0.08191543, 0.92845834, 0.13280773, 0.73471974, 0.76778203,
        0.3412014 , 0.17179807, 0.25780185, 0.42681793, 0.00941415],
       [0.04186662, 0.923118  , 0.23955183, 0.23594396, 0.91821251,
        0.57119709, 0.20157242, 0.50246907, 0.17863503, 0.97733966],
       [0.55526809, 0.18442717, 0.51937208, 0.04620501, 0.60171425,
        0.71572835, 0.99254097, 0.03475022, 0.78310184, 0.60899803],
       [0.22521163, 0.21332454, 0.0234178 , 0.64825087, 0.98136308,
        0.93987854, 0.47655889, 0.51393318, 0.55597511, 0.12123469],
       [0.19024404, 0.30527749, 0.5479686 , 0.79531095, 0.8372735 ,
        0.75507232, 0.58629728, 0.87011028, 0.97745831, 0.55202636],
       [0.01255587, 0.66104685, 0.1597605 , 0.76760107, 0.35154592,
        0.38768925, 0.12878265, 0.31373638, 0.20149068, 0.03631242],
       [0.32786089, 0.60135937, 0.02935237, 0.92580513, 0.45596878,
        0.21332581, 0.57131769, 0.81178661, 0.86553738, 0.36405748],
       [0.61452343, 0.33407375, 0.7128746

8. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as horizontalmente**.

**Dica:** verique as funções **hstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)

In [49]:
a = np.random.randint(0, 101, (10,10))
b = np.random.randint(0, 101, (10,10))
np.hstack((a,b))

array([[54, 89, 63,  5, 45, 28, 52, 99, 22, 63,  1, 54, 39, 62, 59,  7,
         7, 18, 14, 35],
       [43, 22, 43, 64, 97, 74, 48, 40, 18, 18, 50, 52, 71, 65, 14, 82,
        10, 43, 52, 91],
       [26, 85, 86, 34, 98, 25, 74, 71, 74, 52, 32, 80, 84, 96, 64, 43,
         9, 93, 43, 79],
       [67,  9, 84, 48, 60, 50, 82, 57, 59, 34,  3, 43, 33, 49, 22, 98,
        36, 84, 44, 10],
       [81, 80,  9, 75, 90, 52, 28, 53, 55, 71, 39, 72, 16, 25, 61, 60,
        50, 65, 23, 15],
       [10, 46, 15, 58, 77, 99, 84, 55, 75, 13, 76, 87,  9, 26, 43, 10,
        71, 19, 34, 32],
       [23, 67, 14, 49, 60, 30, 90, 67,  8, 31, 41, 32,  5, 80, 15, 71,
        69, 90, 35, 30],
       [60, 64, 23, 50, 64, 46, 30, 71, 12, 11, 46, 38,  0, 65, 26, 64,
        11, 28, 34, 78],
       [84, 10, 53, 60, 26, 77, 31, 38, 50, 94, 52, 10, 39, 14, 91, 14,
        71, 63, 57, 42],
       [41, 22, 54, 68, 98, 36, 47, 64, 25, 49, 41,  5, 51, 91,  3, 35,
         7, 76, 87, 87]])

https://stackoverflow.com/questions/33356442/when-should-i-use-hstack-vstack-vs-append-vs-concatenate-vs-column-stack

9. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as verticalmente**.

Dica: verique as funções **vstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)


In [50]:
a = np.random.randint(0, 101, (10,10))
b = np.random.randint(0, 101, (10,10))
np.vstack((a,b))

array([[ 79,  57,   4,   4,  72,   3,  55,   6,  72,  13],
       [  9,  49,  79,  34,  13,  97,  42,  42,  44,   9],
       [ 93,  34,  45,  61,  70,  70,  81,  53,  42,  68],
       [ 80,  38,  66,  84,  53,  16,   8,  21,  17,  76],
       [ 50,  28,  58,  56,  83,  10,  23,  19,  88,  54],
       [ 65,  63,  86,  17,  49,  33,  77,  84,  84,  39],
       [ 12,   1,  71,   2,  45,  77,  28,  55,  68,  23],
       [ 61,  65,  82,  36,   6,  26,  74,  69,  89,  76],
       [ 55,  16,  95,  34,  86,  75,  57,  58,  88,   1],
       [ 90,  81,   5,  22,  50,  67,  13,  23,  54,  75],
       [ 11,  48,  13,  21,  22,  46,  75,   9,  64,  22],
       [ 43,  87,   5,  61,  60,  59,  35,   8,  67,  48],
       [ 70,  62,   7,  92,  24,  96,   5,   1,  32,  83],
       [ 69,  28,  40,  26,  95,  55,  83,  12,  32,  13],
       [  6,  24,  81,  46,  91,  31,  97,  50,  18,  32],
       [ 94,  59,  52, 100,  32,  79,   5,  48,   3,  88],
       [ 30,  90,  93,  61,  53,  67,  59,  99,   8,  91

Para ler arquivos em formato *txt* que estão em formato de "tabelas", podemos usar a função **genfromtxt()** do **numpy**. Esta função recebe como argumentos: **caminho do arquivo txt** (path local ou uma url); **delimiter** para separar cada valor seguindo algum critério; **dtype** para converter os valores recebidos em algum tipo de dado conhecido (exemplo: **float**) e; **usecols** em que é definido o cabeçalho das colunas lidas a partir do arquivo. 
<br>
<br>
Vejamos um exemplo de uso da função **genfromtxt** sobre a base de dados 
<a href="https://archive.ics.uci.edu/ml/datasets/iris">Iris</a>.

In [ ]:
import numpy as np
# Caminho da base de dados é uma url disponível na internet (visite o link abaixo
# pelo navegador para título de curiosidade :))
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
# Lê os dados a partir do caminho passado e armazena em uma matriz 2D chamada iris_2d
iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

In [ ]:
# matriz 2D criada com os dados da base Iris
iris_2d

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

11. A partir dos dados lidos na matriz iris_2d, filtre as linhas que possuem (coluna 3) > 1.5 e (coluna 1) < 5.0.

12. A partir dos dados lidos na matriz iris_2d, calcule a **média** e **desvio padrão** de cada coluna presente nos dados. Imprima o resultado.

13. Crie um repositório no Github e versione todos os códigos que desenvolveram nesta lista de exercícios.